openvdb_points/tools/IndexFilter.h

```diff
@@ -0,0 +1,495 @@
+///////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2015-2016 Double Negative Visual Effects
+//
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+// Redistributions of source code must retain the above copyright
+// and license notice and the following restrictions and disclaimer.
+//
+// *     Neither the name of Double Negative Visual Effects nor the names
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file IndexFilter.h
+///
+/// @authors Dan Bailey
+///
+/// @brief  Index filters primarily designed to be used with a FilterIndexIter.
+///
+
+
+#ifndef OPENVDB_TOOLS_INDEX_FILTER_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_INDEX_FILTER_HAS_BEEN_INCLUDED
+
+#include <openvdb/version.h>
+#include <openvdb/Types.h>
+
+#include <openvdb/math/Transform.h>
+#include <openvdb/tools/Interpolation.h>
+
+#include <openvdb_points/tools/IndexIterator.h>
+#include <openvdb_points/tools/AttributeArray.h>
+#include <openvdb_points/tools/AttributeGroup.h>
+
+#include <boost/random/uniform_real_distribution.hpp>
+#include <boost/ptr_container/ptr_vector.hpp>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+////////////////////////////////////////
+
+
+namespace index_filter_internal {
+
+template <typename RandGenT, typename IntType>
+struct RandGen
```

Comment:
Am I right in saying that this struct only exists to allow you to pass a `RandInt` into `std::random_shuffle`? We're not currently doing anything with the argument in `operator ()` and yet `std::random_shuffle` expects the return value to be non-negative and less than the value of this argument. It may be worth adding an assert to document this constraint...